File: java-bigdoors/src/main/java/com/bigdoors/domain/ObstructionChecker.java
package com.bigdoors.domain;

import com.bigdoors.util.BlockPos3;
import com.bigdoors.util.Constants;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * Checks whether a door's swing path is obstructed.
 * Pure domain class — no Minecraft imports allowed.
 */
public final class ObstructionChecker {

    private ObstructionChecker() {}

    public enum BlockClassification {
        AIR, SOFT, PASSABLE, SOLID
    }

    public record PathCheckResult(
            boolean canOpen,
            List<BlockPos3> obstructedPositions,
            List<BlockPos3> softBlocks,
            List<BlockPos3> passableBlocks
    ) {}

    public record CloseCheckResult(
            boolean canClose,
            List<BlockPos3> obstructedPositions,
            List<BlockPos3> softBlocks,
            List<BlockPos3> passableBlocks
    ) {}

    public static BlockClassification classifyBlock(String typeId) {
        if (typeId == null || Constants.AIR_BLOCKS.contains(typeId)) {
            return BlockClassification.AIR;
        }
        if (Constants.SOFT_BLOCKS.contains(typeId)) {
            return BlockClassification.SOFT;
        }
        if (Constants.PASSABLE_BLOCKS.contains(typeId)) {
            return BlockClassification.PASSABLE;
        }
        return BlockClassification.SOLID;
    }

    /**
     * Check whether panels can swing in the given direction without hitting solid blocks.
     * Backward-compatible overload defaulting to horizontal mode.
     */
    public static PathCheckResult checkPath(
            List<BlockPos3> panelPositions,
            BlockPos3 hingePos,
            String direction,
            Function<BlockPos3, String> blockQueryFn
    ) {
        return checkPath(panelPositions, hingePos, direction, blockQueryFn, "horizontal", "");
    }

    /**
     * Mode-aware path checking for door rotation.
     */
    public static PathCheckResult checkPath(
            List<BlockPos3> panelPositions,
            BlockPos3 hingePos,
            String direction,
            Function<BlockPos3, String> blockQueryFn,
            String mode,
            String facing
    ) {
        BiFunction<BlockPos3, BlockPos3, BlockPos3> rotateFn =
                RotationMath.getRotateFn(mode, facing, direction);

        Set<String> currentPosKeys = new HashSet<>();
        for (BlockPos3 pos : panelPositions) {
            currentPosKeys.add(pos.toKey());
        }

        List<BlockPos3> obstructed = new ArrayList<>();
        List<BlockPos3> soft = new ArrayList<>();
        List<BlockPos3> passable = new ArrayList<>();

        for (BlockPos3 pos : panelPositions) {
            BlockPos3 dest = rotateFn.apply(pos, hingePos);

            if (currentPosKeys.contains(dest.toKey())) {
                continue;
            }

            String typeId = blockQueryFn.apply(dest);
            BlockClassification classification = classifyBlock(typeId);

            switch (classification) {
                case SOLID -> obstructed.add(dest);
                case SOFT -> soft.add(dest);
                case PASSABLE -> passable.add(dest);
                case AIR -> { /* nothing */ }
            }
        }

        return new PathCheckResult(obstructed.isEmpty(), obstructed, soft, passable);
    }

    /**
     * Check whether a door can close without hitting solid blocks at its closed positions.
     *
     * @param closedPositions    the positions panels will return to
     * @param currentPositionKeys keys of positions currently occupied by panels (being vacated)
     * @param blockQueryFn       function returning typeId at a position
     */
    public static CloseCheckResult checkClose(
            List<BlockPos3> closedPositions,
            Set<String> currentPositionKeys,
            Function<BlockPos3, String> blockQueryFn
    ) {
        List<BlockPos3> obstructed = new ArrayList<>();
        List<BlockPos3> soft = new ArrayList<>();
        List<BlockPos3> passable = new ArrayList<>();

        for (BlockPos3 dest : closedPositions) {
            if (currentPositionKeys.contains(dest.toKey())) {
                continue;
            }

            String typeId = blockQueryFn.apply(dest);
            BlockClassification classification = classifyBlock(typeId);

            switch (classification) {
                case SOLID -> obstructed.add(dest);
                case SOFT -> soft.add(dest);
                case PASSABLE -> passable.add(dest);
                case AIR -> { /* nothing */ }
            }
        }

        return new CloseCheckResult(obstructed.isEmpty(), obstructed, soft, passable);
    }
}

File: bigdoors_bp/scripts/domain/ObstructionChecker.js
/**
 * Obstruction classification and path checking for door rotation.
 *
 * Domain layer — no @minecraft/server imports.
 * Block queries are injected via callback so the domain stays Bedrock-free.
 */

import { SOFT_BLOCKS, PASSABLE_BLOCKS, AIR_BLOCKS } from "../util/Constants.js";
import { getRotateFn } from "./RotationMath.js";

/**
 * Classify a block typeId into one of three tiers.
 * @param {string} typeId — e.g. "minecraft:stone"
 * @returns {'air'|'soft'|'passable'|'solid'}
 */
export function classifyBlock(typeId) {
  if (!typeId || AIR_BLOCKS.has(typeId)) return "air";
  if (SOFT_BLOCKS.has(typeId)) return "soft";
  if (PASSABLE_BLOCKS.has(typeId)) return "passable";
  return "solid";
}

/**
 * Check whether a door can swing in the given direction without hitting solids.
 *
 * @param {Array<{x:number,y:number,z:number}>} panelPositions — current positions of all panels
 * @param {{x:number,y:number,z:number}} hingePos — the hinge to rotate around
 * @param {'cw'|'ccw'} direction — rotation direction
 * @param {function({x:number,y:number,z:number}): string|null} blockQueryFn
 *        Called with a position, returns the typeId of the block there (or null/undefined for air).
 * @param {string} [mode="horizontal"] — rotation mode
 * @param {string} [facing=""] — hinge facing (required for vertical mode)
 * @returns {{canOpen:boolean, obstructedPositions:Array, softBlocks:Array, passableBlocks:Array}}
 */
export function checkPath(panelPositions, hingePos, direction, blockQueryFn, mode = "horizontal", facing = "") {
  const rotateFn = getRotateFn(mode, facing, direction);
  const obstructedPositions = [];
  const softBlocks = [];
  const passableBlocks = [];

  // Also track current panel positions so we don't flag them as obstructions
  const currentPosSet = new Set(
    panelPositions.map((p) => `${p.x},${p.y},${p.z}`)
  );

  for (const pos of panelPositions) {
    const dest = rotateFn(pos, hingePos);
    const destKey = `${dest.x},${dest.y},${dest.z}`;

    // If the destination is a current panel position, it will be vacated — not an obstruction
    if (currentPosSet.has(destKey)) continue;

    const typeId = blockQueryFn(dest);
    const classification = classifyBlock(typeId);

    if (classification === "solid") {
      obstructedPositions.push(dest);
    } else if (classification === "soft") {
      softBlocks.push(dest);
    } else if (classification === "passable") {
      passableBlocks.push(dest);
    }
    // 'air' — no action needed
  }

  return {
    canOpen: obstructedPositions.length === 0,
    obstructedPositions,
    softBlocks,
    passableBlocks,
  };
}

export function checkClose(closedPositions, currentPositionSet, blockQueryFn) {
  const obstructedPositions = [];
  const softBlocks = [];
  const passableBlocks = [];

  for (const dest of closedPositions) {
    const destKey = `${dest.x},${dest.y},${dest.z}`;
    if (currentPositionSet.has(destKey)) continue;

    const typeId = blockQueryFn(dest);
    const classification = classifyBlock(typeId);

    if (classification === "solid") {
      obstructedPositions.push(dest);
    } else if (classification === "soft") {
      softBlocks.push(dest);
    } else if (classification === "passable") {
      passableBlocks.push(dest);
    }
  }

  return {
    canClose: obstructedPositions.length === 0,
    obstructedPositions,
    softBlocks,
    passableBlocks,
  };
}
